File: source/go/internal/astronomy/risesetcache.go
package astronomy

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/store"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const maxCacheEntries = 20_000

var eventCache = store.New[riseSetKey, []float64](maxCacheEntries, store.DefaultStripes, hashRiseSetKey)

// riseSetKey holds the parts of the TypeScript's rise/set cache keys,
// body|direction|lat|lon|elev|dayIndex for events and body|lat|lon|elev|dayIndex
// for scans (direction 0), without formatting five numbers per lookup. Two keys
// are equal exactly when the TypeScript strings are: jsnum.FormatFloat spells
// -0 as 0 and every NaN alike and is otherwise one to one, and jsnum.FormatInt
// spells two integers alike only when they are the same double.
type riseSetKey struct {
	body           RiseSetBody
	direction      int
	lat, lon, elev uint64
	day            uint64
}

func numberKey(v float64) uint64 {
	switch {
	case v == 0:
		return 0
	case v != v:
		return math.Float64bits(math.NaN())
	}
	return math.Float64bits(v)
}

func newRiseSetKey(body RiseSetBody, direction int, location types.GeoLocation, dayIndex int64) riseSetKey {
	return riseSetKey{
		body:      body,
		direction: direction,
		lat:       numberKey(location.Latitude),
		lon:       numberKey(location.Longitude),
		elev:      numberKey(location.Elevation),
		day:       numberKey(float64(dayIndex)),
	}
}

func hashRiseSetKey(k riseSetKey) uint64 {
	h := store.HashString(string(k.body)) ^ uint64(k.direction)
	for _, v := range [...]uint64{k.lat, k.lon, k.elev, k.day} {
		h = store.HashMix(h, v)
	}
	return h
}

type RiseSetKind struct {
	Body RiseSetBody
}

func canonicalDayEventsShared(
	ctx *EphemerisCtx, kind RiseSetKind, direction int,
	location types.GeoLocation, dayIndex int64,
) []float64 {
	key := newRiseSetKey(kind.Body, direction, location, dayIndex)
	events, _ := eventCache.GetOrBuild(key, func() []float64 {
		return append([]float64(nil), dayEventsShared(ctx, kind.Body, direction, location, dayIndex)...)
	})
	return events
}

func clearRiseSetEventCache() {
	eventCache.Clear()
}

func CanonicalDayEvents(
	ctx *EphemerisCtx, kind RiseSetKind, direction int,
	location types.GeoLocation, dayIndex int64,
) []float64 {
	return append([]float64(nil), canonicalDayEventsShared(ctx, kind, direction, location, dayIndex)...)
}

// maxSolvableMs is 2^52: past it a float64 cannot split a millisecond, so the
// root refinement could never reach its tolerance.
const maxSolvableMs = 1 << 52

// ResolveEvent reports [types.ErrInvalidDate] before scanning a day that
// reaches 2^52 ms from 1970, where the solver would never return.
func ResolveEvent(
	ctx *EphemerisCtx, kind RiseSetKind, direction int,
	searchFromMs int64, location types.GeoLocation, limitDays int,
) (int64, bool, error) {
	fromMs := float64(searchFromMs)
	limitMs := fromMs + float64(float64(limitDays)*dayMS)
	startDay := floorDivInt(searchFromMs, dayMS)

	for i := 0; i <= limitDays+1; i++ {
		day := float64(startDay + int64(i))
		if day*dayMS < -maxSolvableMs || (day+1)*dayMS > maxSolvableMs {
			return 0, false, types.Codef(types.ErrInvalidDate,
				"Date must be within 2^52 ms of 1970 for the rise and set solver, got %s",
				types.JSDate(searchFromMs).ISOString())
		}
		for _, event := range canonicalDayEventsShared(ctx, kind, direction, location, startDay+int64(i)) {
			if event >= fromMs {
				if event <= limitMs {
					return int64(event), true, nil
				}
				return 0, false, nil
			}
		}
	}
	return 0, false, nil
}

File: source/ts/src/astronomy/riseSetCache.ts
import { dayEvents, type RiseSetBody } from './riseSet';
import { PanchangError } from '../types/errors';
import { validateDate } from '../utils/validation';
import type { GeoLocation } from '../types/location';

const DAY_MS = 86_400_000;

/** Past 2^52 ms from 1970 a float64 cannot split a millisecond, so the root refinement could never reach its tolerance. */
const MAX_SOLVABLE_MS = 2 ** 52;

/** Keyed per UTC day, always searched from that day's 00:00 UTC. The caller's search
 * start is deliberately *not* in the key: a solver refines to its own tolerance, so
 * keying on it would give one event several timestamps depending on who asked. */
type EventCache = Map<string, readonly number[]>;
const EVENT_CACHE: EventCache = new Map();
const MAX_CACHE_ENTRIES = 20_000;

export interface RiseSetKind {
  body: RiseSetBody;
}

export function canonicalDayEvents(
  kind: RiseSetKind,
  direction: 1 | -1,
  location: GeoLocation,
  dayIndex: number,
): readonly number[] {
  const elevation = location.elevation ?? 0;
  const key =
    `${kind.body}|${direction}|${location.latitude}|${location.longitude}|${elevation}|${dayIndex}`;
  const cached = EVENT_CACHE.get(key);
  if (cached !== undefined) return cached;

  const events = dayEvents(kind.body, direction, location, dayIndex);

  if (EVENT_CACHE.size >= MAX_CACHE_ENTRIES) EVENT_CACHE.clear();
  EVENT_CACHE.set(key, events);
  return events;
}

/** `clearRiseSetTracks` must keep calling this. */
export function clearRiseSetEventCache(): void {
  EVENT_CACHE.clear();
}

/** Throws `INVALID_DATE` for an Invalid Date, or before scanning a day that reaches 2^52 ms from 1970. */
export function resolveEvent(
  kind: RiseSetKind,
  direction: 1 | -1,
  searchFromUtc: Date,
  location: GeoLocation,
  limitDays: number,
): Date | null {
  validateDate(searchFromUtc, 'any');
  const fromMs = searchFromUtc.getTime();
  const limitMs = fromMs + limitDays * DAY_MS;
  const startDay = Math.floor(fromMs / DAY_MS);

  for (let i = 0; i <= limitDays + 1; i++) {
    const day = startDay + i;
    if (day * DAY_MS < -MAX_SOLVABLE_MS || (day + 1) * DAY_MS > MAX_SOLVABLE_MS) {
      throw new PanchangError(
        `Date must be within 2^52 ms of 1970 for the rise and set solver, got ${searchFromUtc.toISOString()}`,
        'INVALID_DATE',
      );
    }
    for (const event of canonicalDayEvents(kind, direction, location, day)) {
      if (event >= fromMs) return event <= limitMs ? new Date(event) : null;
    }
  }
  return null;
}

